<!-- dx-header -->
# MAF Stitcher (DNAnexus Platform App)

Stitch whole-genome alignment blocks in MAF format to generate contiguous
alignments of desired genomic regions, such as transcripts.

This is the source code for an app that runs on the DNAnexus Platform.
For more information about how to run or modify it, see
http://wiki.dnanexus.com/.
<!-- /dx-header -->



<!--
TODO: This app directory was automatically generated by dx-app-wizard;
please edit this Readme.md file to include essential documentation about
your app that would be helpful to users. (Also see the
Readme.developer.md.) Once you're done, you can remove these TODO
comments.

For more info, see http://wiki.dnanexus.com/Developer-Portal.
-->

<!--
TODO: Fill in additional info about how to use each input and output
below.
-->

## Inputs

* **Species set** ``species_set``: ``string`` determines which reference
genome and informant species to include in the output GTable (detailed
below).
* **Desired regions** ``regions``: ``gtable`` a Genes or gri table. The
associated reference genome should match the one in species_set
* **Transcript element type** ``transcript_type``: ``string`` for Genes
tables, specifies which element type to treat as a top-level transcript
(default ``transcript``)
* **Exon element type** ``exon_type``: ``string`` for Genes tables,
specifies which element type to treat as an individual exon to be stitched
into the alignment of a complete transcript (default ``exon``; sensible
to set to ``CDS`` in some contexts)

## Species sets

Human (hg19) centric, from the [UCSC 46-way alignment](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/multiz46way/):

* ``hg19_29mammals``: the 29 mammals analyzed by [Lindblad-Toh et al. (2011)](http://www.nature.com/nature/journal/v478/n7370/full/nature10530.html)
* ``hg19_33mammals``: 29 mammals and four additional primates
* ``hg19_40tetrapods``: human out to *X. tropicalis*
* ``hg19_46vertebrates``: all 46 vertebrate genomes in the MAF alignment  

Human (hg19) centric, from the [UCSC 100-way alignment](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/multiz100way/): 

* ``hg19_58eutherians``: out to & including Armadillo
* ``hg19_62mammals``: out to & including Platypus
* ``hg19_83tetrapods``: out to & including X. tropicalis
* ``hg19_100vertebrates``

*D. melanogaster* (dm3) centric, from the [UCSC 15-way alignment](http://hgdownload.soe.ucsc.edu/goldenPath/dm3/multiz15way/):

* ``dm3_12flies``: [12 *Drosophila* Genomes](http://www.nature.com/nature/journal/v450/n7167/full/nature06340.html)
* ``dm3_15insects``: all 15 insect genomes in the MAF alignment

## Outputs

* **alignments** ``alignments``: ``gtable``

## Output format

The output GTable will have type ``CrossSpeciesAlignments``. The first column,
``_name``, contains the name of a region/transcript from the input regions
GTable. The second column, ``_loci``, will contain a string representation of
the individual genomic loci whose alignments were stitched together for that
region. There will then be one column per species in the requested set, each
containing the gapped sequence for the respective species (the string contents
of each of these columns will all share the same length).

Example:

<pre>
┌───┬───────────┬─────────────────────┬────────────────────────────┬─────────────────
│Row│_name      │_loci                │Human                       │Chimp            
├───┼───────────┼─────────────────────┼────────────────────────────┼─────────────────
│0  │CCDS13560.2│chr21:10973722-10973…│ATGAATGAAAGTCCTGATCCGACTGAC…│ATGAATGAAAGTCCTGA
│1  │CCDS13562.1│chr21:14982549-14983…│ATGGTG-GCTGAGGTTTGTTCAATGCC…│ATGGTG-GCTGAGGTTT
│2  │CCDS13564.1│chr21:15579135-15579…│ATGTTGCTCAAATGTTTAC--ATAATA…│ATGTTGCTCAAATGTTT
│3  │CCDS13567.1│chr21:15755415-15755…│ATGGCCAGAGAGATGACGATCTTAGGA…│ATGGCCAGAGAGATGAC
│4  │CCDS13568.1│chr21:16337039-16340…│ATGACTCATGGAGAAGAGCTTGGCTCT…│ATGACTCATGGAGAAGA
│5  │CCDS13569.1│chr21:18981289-18981…│A--TGAAG--AATGAAA-TTGCTGCCG…│A--TGAAG--AATGAAA
│6  │CCDS13570.1│chr21:19617540-19617…│ATGAGCCGCGTGGTCTCGC-TGCTGCT…│ATGAGCCGCGTGGTCTC
│7  │CCDS13571.1│chr21:19775794-19775…│A-TGGGGTCGAAAAGAGGCATATCTT-…│A-TGGGGTCAAAAAGAA
│8  │CCDS13573.1│chr21:26979714-26979…│ATGGAGG----CGCTGGCCATGG----…│ATGGAGG----CGCTGG
│9  │CCDS13574.1│chr21:27101941-27102…│ATGATTCTTCAGAGGCTCTTCAGGTTC…│ATGATTCTTCAGAGGCT
└───┴───────────┴─────────────────────┴────────────────────────────┴─────────────────
</pre>


